lib/k8s/client/websocket_provider.ex

```diff
@@ -0,0 +1,97 @@
+defmodule K8s.Client.WebSocketProvider do
+  @moduledoc """
+  Websocket client for k8s API.
+  """
+  require Logger
+  use WebSockex
+
+  @doc """
+  Make a connection to the K8s API and upgrade to a websocket. This is useful for streaming resources like /exec, /logs, /attach  
+
+  ## Example
+
+  ```elixir
+  url = "https://localhost:6443/api/v1/namespaces/test/pods/nginx-pod/exec?command=%2Fbin%2Fsh&command=-c&command=date&stdin=true&stderr=true&stdout=true&tty=true"
+  insecure = false
+  ssl_options = [cert: <<1,2>>, key: {:RSAPrivateKey, <<48, 130>>}, verify: :verify_none]
+  cacerts = [<<1,2,3>>,<1,2,3>>]
+  headers = [{"Accept", "*/*"}, {"Content-Type", "application/json"}]
+  opts = [command    : ["/bin/sh", "-c", "date"], stdin: true, stderr: true, stdout: true, tty: true, stream_to: #PID<0.271.0>]
+  {:ok, pid} = K8s.Client.WebSocketProvider.request(url, ssl_options, cacerts, headers, opts)
+  ```
+
+  """
+  @spec request(
+          String.t(),
+          boolean(),
+          keyword(atom()),
+          keyword(binary()),
+          keyword(tuple()),
+          keyword(atom())
+        ) ::
+          {:ok, pid()} | {:error, binary()}
+  def request(url, insecure, ssl_options, cacerts, headers, opts) do
+    conn =
+      WebSockex.Conn.new(url,
+        insecure: insecure,
+        ssl_options: ssl_options,
+        cacerts: cacerts,
+        extra_headers: headers
+      )
+
+    WebSockex.start_link(conn, __MODULE__, opts, async: true)
```

Comment:
The `stream_to` pid should be monitored. if `stream_to` != caller you'd end up with a zombie process in case `stream_to` dies.
Since `websocketex` doesn't seem to provide a `init` callback, you should probably wrap ws in some process or perhaps setup the monitor in `handle_connect/2`?

```elixir
  defstruct [:monitor_ref, :stream_to]

  def request(url, insecure, ssl_options, cacerts, headers, opts) do
    conn =
      WebSockex.Conn.new(url,
        insecure: insecure,
        ssl_options: ssl_options,
        cacerts: cacerts,
        extra_headers: headers
      )

    state = %__MODULE__{stream_to: opts[:stream_to]}
    WebSockex.start_link(conn, __MODULE__, state, async: true)
  end

  def handle_connect(_conn, %{monitor_ref: nil, stream_to: stream_to} = state) do
    ref = Process.monitor(stream_to)

    {:ok, %{state | monitor_ref: ref}}
  end

  # In case of a disconnects handle_connect/2 will be called again
  def handle_connect(_conn, state) do
    {:ok, state}
  end

  def handle_info({:DOWN, ref, :process, _, _}, %{monitor_ref: ref}) do
    {:"$EXIT", {:stop, :normal}}
  end

```

Reply:
indeed! Very nice, thanks for this!

Reply:
I ignored the warn from dialyzer on `handle_info/2` as the behavior in Websockex is a bit different.